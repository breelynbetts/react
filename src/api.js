const apiHost = () => {} // No-op in our mock version.

const searchNasa = () => Promise.resolve({
  data: [{
      "id": 689597,
      "sol": 2540,
      "camera": {
        "id": 20,
        "name": "FHAZ",
        "rover_id": 5,
        "full_name": "Front Hazard Avoidance Camera"
      },
      "img_src": "https://mars.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/02540/opgs/edr/fcam/FLB_622990123EDR_F0763002FHAZ00341M_.JPG",
      "earth_date": "2019-09-28",
      "rover": {
        "id": 5,
        "name": "Curiosity",
        "landing_date": "2012-08-06",
        "launch_date": "2011-11-26",
        "status": "active",
        "max_sol": 2540,
        "max_date": "2019-09-28",
        "total_photos": 366206,
        "cameras": [{
            "name": "FHAZ",
            "full_name": "Front Hazard Avoidance Camera"
          },
          {
            "name": "NAVCAM",
            "full_name": "Navigation Camera"
          },
          {
            "name": "MAST",
            "full_name": "Mast Camera"
          },
          {
            "name": "CHEMCAM",
            "full_name": "Chemistry and Camera Complex"
          },
          {
            "name": "MAHLI",
            "full_name": "Mars Hand Lens Imager"
          },
          {
            "name": "MARDI",
            "full_name": "Mars Descent Imager"
          },
          {
            "name": "RHAZ",
            "full_name": "Rear Hazard Avoidance Camera"
          }
        ]
      }
    },
    {
      "id": 689598,
      "sol": 2540,
      "camera": {
        "id": 20,
        "name": "FHAZ",
        "rover_id": 5,
        "full_name": "Front Hazard Avoidance Camera"
      },
      "img_src": "https://mars.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/02540/opgs/edr/fcam/FRB_622990123EDR_F0763002FHAZ00341M_.JPG",
      "earth_date": "2019-09-28",
      "rover": {
        "id": 5,
        "name": "Curiosity",
        "landing_date": "2012-08-06",
        "launch_date": "2011-11-26",
        "status": "active",
        "max_sol": 2540,
        "max_date": "2019-09-28",
        "total_photos": 366206,
        "cameras": [{
            "name": "FHAZ",
            "full_name": "Front Hazard Avoidance Camera"
          },
          {
            "name": "NAVCAM",
            "full_name": "Navigation Camera"
          },
          {
            "name": "MAST",
            "full_name": "Mast Camera"
          },
          {
            "name": "CHEMCAM",
            "full_name": "Chemistry and Camera Complex"
          },
          {
            "name": "MAHLI",
            "full_name": "Mars Hand Lens Imager"
          },
          {
            "name": "MARDI",
            "full_name": "Mars Descent Imager"
          },
          {
            "name": "RHAZ",
            "full_name": "Rear Hazard Avoidance Camera"
          }
        ]
      }
    },
    {
      "id": 689599,
      "sol": 2540,
      "camera": {
        "id": 21,
        "name": "RHAZ",
        "rover_id": 5,
        "full_name": "Rear Hazard Avoidance Camera"
      },
      "img_src": "https://mars.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/02540/opgs/edr/rcam/RLB_622990164EDR_F0763002RHAZ00341M_.JPG",
      "earth_date": "2019-09-28",
      "rover": {
        "id": 5,
        "name": "Curiosity",
        "landing_date": "2012-08-06",
        "launch_date": "2011-11-26",
        "status": "active",
        "max_sol": 2540,
        "max_date": "2019-09-28",
        "total_photos": 366206,
        "cameras": [{
            "name": "FHAZ",
            "full_name": "Front Hazard Avoidance Camera"
          },
          {
            "name": "NAVCAM",
            "full_name": "Navigation Camera"
          },
          {
            "name": "MAST",
            "full_name": "Mast Camera"
          },
          {
            "name": "CHEMCAM",
            "full_name": "Chemistry and Camera Complex"
          },
          {
            "name": "MAHLI",
            "full_name": "Mars Hand Lens Imager"
          },
          {
            "name": "MARDI",
            "full_name": "Mars Descent Imager"
          },
          {
            "name": "RHAZ",
            "full_name": "Rear Hazard Avoidance Camera"
          }
        ]
      }
    }
  ]
})

const searchAstronomyPicture = () => Promise.resolve({
  data: [{
    "copyright": "Eric Wagner",
    "date": "2019-10-10",
    "explanation": "On September 24, a late evening commercial flight from Singapore to Australia offered stratospheric views of the southern hemisphere's night sky, if you chose a window seat. In fact, a well-planned seating choice with a window facing toward the Milky Way allowed the set up of a sensitive digital camera on a tripod mount to record the galaxy's central bulge in a series of 10 second long exposures.  By chance, one of the exposures caught this bright fireball meteor in the starry frame. Reflected along the wing of the A380 aircraft, the brilliant greenish streak is also internally reflected in the double layer window, producing a fainter parallel to the original meteor track. In the southern sky Jupiter is the bright source beneath the galactic bulge and seen next to a green beacon, just off the wing tip.",
    "hdurl": "https://apod.nasa.gov/apod/image/1910/MWBolideEricWagner2400.jpg",
    "media_type": "image",
    "service_version": "v1",
    "title": "Mid-Air Meteor and Milky Way",
    "url": "https://apod.nasa.gov/apod/image/1910/MWBolideEricWagner1200.jpg"
  }]

})

const searchEvents = () => Promise.resolve({
  data: [{
    "title": "EONET Events",
    "description": "Natural events from EONET.",
    "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events",
    "events": [
      {
        "id": "EONET_4475",
        "title": "Wildfires - Los Angeles County (Tick Fire), California, United States",
        "description": "",
        "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events/EONET_4475",
        "categories": [
          {
            "id": 8,
            "title": "Wildfires"
          }
        ],
        "sources": [
          {
            "id": "PDC",
            "url": "http://emops.pdc.org/emops/?hazard_id=96981"
          }
        ],
        "geometries": [
          {
            "date": "2019-10-24T18:16:00Z",
            "type": "Point",
            "coordinates": [
              -118.39468413,
              34.457634347
            ]
          }
        ]
      },
      {
        "id": "EONET_4476",
        "title": "Wildfires - San Bernardino County (Old Water Fire), California, United States",
        "description": "",
        "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events/EONET_4476",
        "categories": [
          {
            "id": 8,
            "title": "Wildfires"
          }
        ],
        "sources": [
          {
            "id": "PDC",
            "url": "http://emops.pdc.org/emops/?hazard_id=96977"
          }
        ],
        "geometries": [
          {
            "date": "2019-10-24T15:11:00Z",
            "type": "Point",
            "coordinates": [
              -117.26913,
              34.17922
            ]
          }
        ]
      },
      {
        "id": "EONET_4474",
        "title": "Wildfire -Sonoma County (Kincade Fire), California, United States",
        "description": "",
        "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events/EONET_4474",
        "categories": [
          {
            "id": 8,
            "title": "Wildfires"
          }
        ],
        "sources": [
          {
            "id": "PDC",
            "url": "http://emops.pdc.org/emops/?hazard_id=96971"
          }
        ],
        "geometries": [
          {
            "date": "2019-10-24T13:32:00Z",
            "type": "Point",
            "coordinates": [
              -122.807158563,
              38.758962281
            ]
          }
        ]
      }
    ]
  }  
  ]

})

export {
  apiHost,
  searchNasa,
  searchAstronomyPicture,
  searchEvents
}
