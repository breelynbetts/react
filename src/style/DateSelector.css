form {
    text-align: center;
    width: 100%;
    align-items: center;
    justify-content: center;
    display: inline;
}

input {
    display: inline;
}
button {
    background-color: white;
    border: 2px solid #008CBA;
}

button:hover  {
    background-color: #008CBA;
    color:black;
} 

button:disabled,
button[disabled] {
    border: 1px solid #999999;
    background-color: #cccccc;
     color: #666666;
}