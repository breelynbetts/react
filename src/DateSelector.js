import React, { useState } from 'react'
import { searchAstronomyPicture } from './api.js'
import DateResult from './DateResult.js'

import './style/DateSelector.css'

function DateSelector() {
  const [error, setError] = useState(null)
  const [query, setQuery] = useState('')
  const [date, setDate] = useState([])
  
  const handleQueryChange = event => setQuery(event.target.value)
  
  const handleInput = async event  => {
      event.preventDefault()
  
      setError(null)
  
      try {
        const result = await searchAstronomyPicture({
          q: query
        })
  
        setDate(result.data)
      } catch (error) {
        setError('Sorry, but something went wrong.')
      }
  }

  return (
      <div>
        <form className="DatePicker" onSubmit={handleInput}>
          <input name="query" type="date" value={query} onChange={handleQueryChange}/>
          <div className="ButtonDiv">
            <button type="submit" disabled={!query}>Select Date</button>
          </div>
          {error && (
            <div className="error">
              {error}
              <p>Error has occurred!</p>
            </div>
          )}
        </form>
        <p>Date Selected: {query}</p>
        <DateResult results={date}/>
    </div>
  )
}
export default DateSelector