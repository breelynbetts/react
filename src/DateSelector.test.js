import React from 'react'
import ReactDOM from 'react-dom'
import TestRenderer from 'react-test-renderer'
import ReactTestUtils from 'react-dom/test-utils'

import sinon from 'sinon'

import DateSelector from './DateSelector'

import * as api from './api'

it('should start with no date selected', () => {
    const component = TestRenderer.create(<DateSelector />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

it('should start with a disabled selection button', () => {
    const component = TestRenderer.create(<DateSelector />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})

describe('date button', () => {
    let div
    beforeEach(() => {
        div = document.createElement('div')
        ReactTestUtils.act(() => {
            ReactDOM.render(<DateSelector />, div)
        })
    })

    afterEach(() => ReactDOM.unmountComponentAtNode(div))

    it('should be enabled when the date field is fully filled out', () => {
        const dateInput = div.querySelector('input')
        ReactTestUtils.act(() => {
            dateInput.value = '1999-09-25'
            ReactTestUtils.Simulate.change(dateInput)
        })

        const dateButton = div.querySelector('button')
        expect(dateButton.disabled).toBe(false)
    })

    it('should be disabled when a valid date is not entered', () => {
        const dateInput = div.querySelector('input')
        ReactTestUtils.act(() => {
            dateInput.value = 'this should break!'
            ReactTestUtils.Simulate.change(dateInput)
        })

        const dateButton = div.querySelector('button')
        expect(dateButton.disabled).toBe(true)
    })
})
  

const setupAndQueryDataForm = async () => {
    const div = document.createElement('div')
    ReactTestUtils.act(() => {
      ReactDOM.render(<DateSelector />, div)
    })
  
    const dateInput = div.querySelector('input')
    ReactTestUtils.act(() => {
      dateInput.value = '1999-09-25'
      ReactTestUtils.Simulate.change(dateInput)
    })
  
    const dateSelector = div.querySelector('form')
    await ReactTestUtils.act(async () => {
      await ReactTestUtils.Simulate.submit(dateSelector)
    })
  
    return div
}

describe('API calls', () => {
    let div
    beforeEach(async () => {
      sinon.stub(api, 'searchAstronomyPicture')
  
      // To manage size, we supply a mock response that contains _only_ what the app will need. This does mean
      // that we need to revise the mock response if our app starts using more (or different) data.
      api.searchAstronomyPicture.returns(Promise.resolve({
        data: [{
            "copyright": "Eric Wagner",
            "date": "2019-10-10",
            "explanation": "On September 24, a late evening commercial flight from Singapore to Australia offered stratospheric views of the southern hemisphere's night sky, if you chose a window seat. In fact, a well-planned seating choice with a window facing toward the Milky Way allowed the set up of a sensitive digital camera on a tripod mount to record the galaxy's central bulge in a series of 10 second long exposures.  By chance, one of the exposures caught this bright fireball meteor in the starry frame. Reflected along the wing of the A380 aircraft, the brilliant greenish streak is also internally reflected in the double layer window, producing a fainter parallel to the original meteor track. In the southern sky Jupiter is the bright source beneath the galactic bulge and seen next to a green beacon, just off the wing tip.",
            "hdurl": "https://apod.nasa.gov/apod/image/1910/MWBolideEricWagner2400.jpg",
            "media_type": "image",
            "service_version": "v1",
            "title": "Mid-Air Meteor and Milky Way",
            "url": "https://apod.nasa.gov/apod/image/1910/MWBolideEricWagner1200.jpg"
        }]
      }))
  
      div = await setupAndQueryDataForm()
    })
  
    afterEach(() => {
      ReactDOM.unmountComponentAtNode(div)
      api.searchAstronomyPicture.restore()
    })
  
    it('should trigger a Astronomy Picture search when the search button is clicked', () => {
      // Note how this _isn’t_ a snapshot test because we’re checking whether a function was called with
      // the right arguments.
      expect(api.searchAstronomyPicture.firstCall.args[0]).toEqual({
        date: '1999-09-25' // Our test search term.
      })
    })
  
    it('should populate the image container when search results arrive', () => {
      // Our mock search results yield one image, so we expect our results container to have one child.
      const searchResults = div.querySelector('div.DateResult')
      expect(searchResults.children.length).toEqual(1)
    })
})

describe('failed API calls', () => {
    let div
    beforeEach(async () => {
      sinon.stub(api, 'searchAstronomyPicture')
      api.searchAstronomyPicture.returns(Promise.reject('Mock failure'))
  
      div = await setupAndQueryDataForm()
    })
  
    afterEach(() => {
      ReactDOM.unmountComponentAtNode(div)
      api.searchAstronomyPicture.restore()
    })
  
    it('should display an alert when the API call fails', () => {
      // The document should contain the error div.
      const searchError = div.querySelector('div.error')
      expect(searchError.textContent).toEqual('Sorry, but something went wrong.')
    })
  })