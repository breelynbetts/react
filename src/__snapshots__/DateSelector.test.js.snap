// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should start with a disabled selection button 1`] = `
<div>
  <form
    className="DatePicker"
    onSubmit={[Function]}
  >
    <input
      max="2019-11-14"
      min="1995-06-21"
      name="query"
      onChange={[Function]}
      type="date"
      value=""
    />
    <div
      className="ButtonDiv"
    >
      <button
        disabled={true}
        type="submit"
      >
        select date
      </button>
    </div>
  </form>
  <div
    className="DateResult"
  >
    <div
      className="DateResult"
    />
  </div>
</div>
`;

exports[`should start with no date selected 1`] = `
<div>
  <form
    className="DatePicker"
    onSubmit={[Function]}
  >
    <input
      max="2019-11-14"
      min="1995-06-21"
      name="query"
      onChange={[Function]}
      type="date"
      value=""
    />
    <div
      className="ButtonDiv"
    >
      <button
        disabled={true}
        type="submit"
      >
        select date
      </button>
    </div>
  </form>
  <div
    className="DateResult"
  >
    <div
      className="DateResult"
    />
  </div>
</div>
`;
